ACLiC compilation confuses compiled binaries with shared objects, breaking compilation in some cases

- [X] Checked for duplicates
<!--
Please search in
* [GitHub](https://github.com/root-project/root/issues?q=is%3Aissue)
* AND [Jira](https://sft.its.cern.ch/jira/issues/?jql=project %3D ROOT)
for existing reports of your issue.

If you find one, you are very welcome to add to the existing report, for instance "issue still exists in today's master".
-->

### Describe the bug
<!--
A clear and concise description of what the wrong behavior is.
-->

Building ROOT on Arch Linux and then running `ctest -R "(dataframe|datasource)" -- -j8` results in several test failures.
There are various possible failure modes -- the command above typically results in these kind of errors:

```
1369: Processing /home/jalopezg/CERN/roottest/root/dataframe/test_snapshot.C+...
1369: Info in <TUnixSystem::ACLiC>: creating shared library /home/jalopezg/CERN/build/roottest/root/dataframe/test_snapshot_C.so
1369: /usr/bin/ld: /home/jalopezg/CERN/build/roottest/root/dataframe/test_readFcc: _ZSt4cout: invalid version 6 (max 0)
1369: /usr/bin/ld: /home/jalopezg/CERN/build/roottest/root/dataframe/test_readFcc: error adding symbols: bad value
1369: collect2: error: ld returned 1 exit status
1369: Error in <ACLiC>: Executing 'cd "/home/jalopezg/CERN/build/roottest/root/dataframe" ; c++ -fPIC -c -O3 -DNDEBUG -std=c++14 -Wno-implicit-fallthrough -Wno-noexcept-type -pipe -W -Woverloaded-virtual -fsigned-char -pthread  -I$ROOTSYS/include -I/home/jalopezg/CERN/build/roottest/root/dataframe -I"/home/jalopezg/CERN/build/etc/" -I"/home/jalopezg/CERN/build/etc//cling" -I"/home/jalopezg/CERN/build/include/" -I"/home/jalopezg/CERN/build/include" -I"/home/jalopezg/CERN/build/roottest/root/dataframe"   -D__ACLIC__ "/home/jalopezg/CERN/build/roottest/root/dataframe/test_snapshot_C_ACLiC_dict.cxx" ; c++ -O3 -DNDEBUG "/home/jalopezg/CERN/build/roottest/root/dataframe/test_snapshot_C_ACLiC_dict.o" -shared   "/usr/lib/libdl.so" "/usr/lib/libc.so" "/usr/lib/libgcc_s.so" "/usr/lib/libm.so" "/usr/lib/libstdc++.so" "/home/jalopezg/CERN/build/lib/libRint.so" "/home/jalopezg/CERN/build/lib/libCore.so" "/usr/lib/libpthread.so.0" "/usr/lib/libpcre.so.1" "/usr/lib/liblzma.so.5" "/usr/lib/liblz4.so.1" "/usr/lib/libz.so.1" "/usr/lib/libzstd.so.1" "/usr/lib/libnss_files.so.2" "/home/jalopezg/CERN/build/lib/libRIO.so" "/home/jalopezg/CERN/build/lib/libThread.so" "/home/jalopezg/CERN/build/lib/libCling.so" "/usr/lib/librt.so.1" "/usr/lib/libncursesw.so.6" "/usr/lib/ld-2.33.so" "/home/jalopezg/CERN/build/lib/libMathCore.so" "/home/jalopezg/CERN/build/lib/libImt.so" "/home/jalopezg/CERN/build/lib/libMultiProc.so" "/home/jalopezg/CERN/build/lib/libNet.so" "/usr/lib/libtbb.so.2" "/usr/lib/libssl.so.1.1" "/usr/lib/libcrypto.so.1.1" "/usr/lib/libnss_systemd.so.2" -o "/home/jalopezg/CERN/build/roottest/root/dataframe/test_snapshot_C.so" /home/jalopezg/CERN/build/lib/libROOTDataFrame.so /home/jalopezg/CERN/build/lib/libTreePlayer.so /home/jalopezg/CERN/build/lib/libTree.so /home/jalopezg/CERN/build/roottest/root/dataframe/test_readFcc /home/jalopezg/CERN/build/roottest/root/dataframe/par' failed!
1369: terminate called after throwing an instance of 'std::logic_error'
1369:   what():  basic_string::_M_construct null not valid
```

I did not manage to reproduce the problem outside of ctest and without running multiple tests concurrently -- it also seems that one of the tests "has" to be `roottest-root-dataframe-test_snapshot_manytasks`.

### To Reproduce
<!--
Steps to reproduce the behavior:
1. Your code that triggers the issue: at least a part; ideally something we can run ourselves.
2. Don't forget to attach the required input files!
3. How to run your code and / or build it, e.g. `root myMacro.C`, ...
-->

The following repository provides a Dockerfile to reproduce the issue in a container https://gitlab.cern.ch/eguiraud/arch_aclic_bug .

### Setup
<!--
1. ROOT version
2. Operating system
3. How you obtained ROOT, such as `dnf install` / binary download / you built it yourself.
-->

ROOT master (but 6.22 also has issues) and arch linux.

I tried building from source (6.22 and master) as well as installing from the system repos -- failure modes are different but none can run all tests without problems.

Gentoo seems to also be affected.

--- a/interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
+++ b/interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
@@ -590,10 +590,13 @@ namespace cling {
   void HandleDynTab(const ELFFile<ELFT>* Elf, llvm::StringRef FileName,
                     llvm::SmallVector<llvm::StringRef,2>& RPath,
                     llvm::SmallVector<llvm::StringRef,2>& RunPath,
-                    std::vector<StringRef>& Deps) {
+                    std::vector<StringRef>& Deps,
+                    bool& isPIEExecutable) {
     const char *Data = "";
     if (Expected<StringRef> StrTabOrErr = getDynamicStrTab(Elf))
       Data = StrTabOrErr.get().data();
+
+    isPIEExecutable = false;
 
     auto DynamicEntriesOrError = Elf->dynamicEntries();
     if (!DynamicEntriesOrError) {
@@ -612,6 +615,11 @@ namespace cling {
           break;
         case ELF::DT_RUNPATH:
           SplitPaths(Data + Dyn.d_un.d_val, RunPath, utils::kAllowNonExistant, platform::kEnvDelim, false);
+          break;
+        case ELF::DT_FLAGS_1:
+          // Check if this is not a pie executable.
+          if (Dyn.d_un.d_val & llvm::ELF::DF_1_PIE)
+            isPIEExecutable = true;
           break;
         // (Dyn.d_tag == ELF::DT_NULL) continue;
         // (Dyn.d_tag == ELF::DT_AUXILIARY || Dyn.d_tag == ELF::DT_FILTER)
@@ -731,15 +739,28 @@ namespace cling {
           }
           llvm::object::ObjectFile *BinObjF = ObjFileOrErr.get().getBinary();
           if (BinObjF->isELF()) {
-            if (const auto* ELF = dyn_cast<ELF32LEObjectFile>(BinObjF))
-              HandleDynTab(ELF->getELFFile(), FileName, RPath, RunPath, Deps);
-            else if (const auto* ELF = dyn_cast<ELF32BEObjectFile>(BinObjF))
-              HandleDynTab(ELF->getELFFile(), FileName, RPath, RunPath, Deps);
-            else if (const auto* ELF = dyn_cast<ELF64LEObjectFile>(BinObjF))
-              HandleDynTab(ELF->getELFFile(), FileName, RPath, RunPath, Deps);
-            else if (const auto* ELF = dyn_cast<ELF64BEObjectFile>(BinObjF))
-              HandleDynTab(ELF->getELFFile(), FileName, RPath, RunPath, Deps);
+            bool isPIEExecutable = false;
 
+            if (const auto* ELF = dyn_cast<ELF32LEObjectFile>(BinObjF))
+              HandleDynTab(ELF->getELFFile(), FileName, RPath, RunPath, Deps,
+                           isPIEExecutable);
+            else if (const auto* ELF = dyn_cast<ELF32BEObjectFile>(BinObjF))
+              HandleDynTab(ELF->getELFFile(), FileName, RPath, RunPath, Deps,
+                           isPIEExecutable);
+            else if (const auto* ELF = dyn_cast<ELF64LEObjectFile>(BinObjF))
+              HandleDynTab(ELF->getELFFile(), FileName, RPath, RunPath, Deps,
+                           isPIEExecutable);
+            else if (const auto* ELF = dyn_cast<ELF64BEObjectFile>(BinObjF))
+              HandleDynTab(ELF->getELFFile(), FileName, RPath, RunPath, Deps,
+                           isPIEExecutable);
+
+            if ((level == 0) && isPIEExecutable) {
+              if (searchSystemLibraries)
+                m_SysLibraries.UnregisterLib(LibPath);
+              else
+                m_Libraries.UnregisterLib(LibPath);
+              return;
+            }
           } else if (BinObjF->isMachO()) {
             MachOObjectFile *Obj = (MachOObjectFile*)BinObjF;
             for (const auto &Command : Obj->load_commands()) {

--- a/interpreter/cling/test/DynamicLibraryManager/cached_realpath.C
+++ b/interpreter/cling/test/DynamicLibraryManager/cached_realpath.C
@@ -8,6 +8,7 @@
 
 // REQUIRES: not_system-windows
 
+// RUN: %mkdir %t-dir
 // RUN: cd %t-dir
 
 // RUN: %mkdir %t-dir/dir1/dir11/dir111

--- a/interpreter/cling/test/DynamicLibraryManager/call_pie.c
+++ b/interpreter/cling/test/DynamicLibraryManager/call_pie.c
@@ -1,0 +1,17 @@
+/*------------------------------------------------------------------------------
+// CLING - the C++ LLVM-based InterpreterG :)
+//
+// This file is dual-licensed: you can choose to license it under the University
+// of Illinois Open Source License or the GNU Lesser General Public License. See
+// LICENSE.TXT for details.
+//----------------------------------------------------------------------------*/
+
+// RUN: true
+// Used as executable/library source by pie.C, etc.
+CLING_EXPORT int cling_testlibrary_function() {
+  return 42;
+}
+
+int main() {
+  return 0;
+}

--- a/interpreter/cling/test/DynamicLibraryManager/callable_lib_L_AB_order1.C
+++ b/interpreter/cling/test/DynamicLibraryManager/callable_lib_L_AB_order1.C
@@ -26,6 +26,5 @@ cling_testlibrary_function()
 .L libcall_lib_L_AB
 cling_testlibrary_function()
 // CHECK: (int) 91749
-357
 
 .q

--- a/interpreter/cling/test/DynamicLibraryManager/pie.C
+++ b/interpreter/cling/test/DynamicLibraryManager/pie.C
@@ -1,0 +1,39 @@
+//------------------------------------------------------------------------------
+// CLING - the C++ LLVM-based InterpreterG :)
+//
+// This file is dual-licensed: you can choose to license it under the University
+// of Illinois Open Source License or the GNU Lesser General Public License. See
+// LICENSE.TXT for details.
+//------------------------------------------------------------------------------
+
+// REQUIRES: not_system-windows
+
+// RUN: mkdir -p %t-dir/lib
+// RUN: %clang -shared -DCLING_EXPORT=%dllexport %S/call_lib_A.c -o%t-dir/lib/libcall_lib_A%shlibext
+// RUN: %clang -shared -DCLING_EXPORT=%dllexport %S/call_lib_B.c -o%t-dir/lib/libcall_lib_B%shlibext
+// RUN: %clang %fPIC -fpie -pie -DCLING_EXPORT=%dllexport %S/call_pie.c -o%t-dir/lib/call_pie_so%shlibext
+// RUN: %clang %fPIC -fpie -pie -DCLING_EXPORT=%dllexport %S/call_pie.c -o%t-dir/lib/call_pie
+// RUN: cat %s | LD_LIBRARY_PATH="%t-dir/lib" %cling 2>&1 | FileCheck %s
+
+// Test: Cling pragma for loading libraries. Lookup and load library Lib_L_AB
+//       that depends on two libraries Lib_A and Lib_B via file names.
+//       Lib_A and Lib_B are in LD_LIBRARY_PATH. Call_pie is "PIE" compiled excutable.
+//       Lookup functions from libraries and use them to print result value.
+//       We are expecting an error for the symbol from the PIE executable, which is
+//       expected not to be loaded by dynamic linker.
+
+#pragma cling load("libcall_lib_A")
+extern "C" int cling_testlibrary_function_A();
+cling_testlibrary_function_A()
+// CHECK: (int) 170
+
+#pragma cling load("libcall_lib_B")
+extern "C" int cling_testlibrary_function_B();
+cling_testlibrary_function_B()
+// CHECK: (int) 187
+
+extern "C" int cling_testlibrary_function();
+cling_testlibrary_function()
+// expected-error {{symbol 'cling_testlibrary_function' unresolved while linking}}
+
+//.q

--- a/interpreter/llvm/src/include/llvm/BinaryFormat/ELF.h
+++ b/interpreter/llvm/src/include/llvm/BinaryFormat/ELF.h
@@ -1289,7 +1289,8 @@ enum {
   DF_1_NORELOC = 0x00400000,
   DF_1_SYMINTPOSE = 0x00800000, // Object has individual interposers.
   DF_1_GLOBAUDIT = 0x01000000,  // Global auditing required.
-  DF_1_SINGLETON = 0x02000000   // Singleton symbols are used.
+  DF_1_SINGLETON = 0x02000000,  // Singleton symbols are used.
+  DF_1_PIE = 0x08000000,        // Object is a position-independent executable.
 };
 
 // DT_MIPS_FLAGS values.

